File: pt/tsc.cpp
#include "tsc.h"
#include <assert.h>

namespace hase::pt {
TscConverter::TscConverter(uint64_t _timeZero, uint16_t _timeShift,
                           uint32_t _timeMult)
    : timeZero(_timeZero), timeShift(_timeShift), timeMult(_timeMult) {
  assert(timeMult != 0);
}

uint64_t TscConverter::tscToPerfTime(uint64_t tsc) {
  uint64_t quot = (tsc >> timeShift) * timeMult;
  uint64_t rem = ((tsc & ((1ull << timeShift) - 1)) * timeMult) >> timeShift;
  return timeZero + quot + rem;
}

uint64_t TscConverter::perfTimeToTsc(uint64_t time) {
  time -= timeZero;
  return ((time / timeMult) << timeShift) +
         (((time % timeMult) << timeShift) / timeMult);
}
} // namespace hase::pt

File: hase/perf/tsc.py
class TscConversion:
    def __init__(self, time_mult: int, time_shift: int, time_zero: int) -> None:
        self.time_mult = time_mult
        self.time_shift = time_shift
        self.time_zero = time_zero

    def tsc_to_perf_time(self, tsc: int) -> int:
        quot = (tsc >> self.time_shift) * self.time_mult
        rem = ((tsc & ((1 << self.time_shift) - 1)) * self.time_mult) >> self.time_shift
        return self.time_zero + quot + rem

    def perf_time_to_tsc(self, time: int) -> int:
        time -= self.time_zero
        return ((time // self.time_mult) << self.time_shift) + (
            ((time % self.time_mult) << self.time_shift) // self.time_mult
        )


if __name__ == "__main__":
    t = TscConversion(2, 3, 4)
    assert 50 == t.tsc_to_perf_time(t.perf_time_to_tsc(50))
